src/ngraph/opsets/opset.hpp

```diff
@@ -16,35 +16,87 @@
 #include "ngraph/node.hpp"
 
 namespace ngraph
 {
+    /// \brief Run-time opset information
     class OpSet
     {
+        static std::mutex& get_mutex();
+
     public:
-        OpSet(const std::set<NodeTypeInfo>& op_types)
-            : m_op_types(op_types)
+        std::set<NodeTypeInfo>::size_type size() const
+        {
+            std::lock_guard<std::mutex> guard(get_mutex());
+            return m_op_types.size();
+        }
+        /// \brief Insert an op into the opset with a particular name and factory
+        void insert(const std::string& name,
+                    const NodeTypeInfo& type_info,
+                    FactoryRegistry<Node>::Factory factory)
+        {
+            std::lock_guard<std::mutex> guard(get_mutex());
+            m_op_types.insert(type_info);
+            m_name_type_info_map[name] = type_info;
+            ngraph::FactoryRegistry<Node>::get().register_factory(type_info, factory);
+        }
+
+        /// \brief Insert OP_TYPE into the opset with a special name and the default factory
+        template <typename OP_TYPE>
+        void insert(const std::string& name)
+        {
+            insert(name, OP_TYPE::type_info, FactoryRegistry<Node>::get_default_factory<OP_TYPE>());
+        }
+
+        /// \brief Insert OP_TYPE into the opset with the default name and factory
+        template <typename OP_TYPE>
+        void insert()
+        {
+            insert<OP_TYPE>(OP_TYPE::type_info.name);
+        }
+
+        /// \brief Create the op named name using it's factory
+        ngraph::Node* create(const std::string& name) const;
+
+        /// \brief Return true if OP_TYPE is in the opset
+        bool contains_type(const NodeTypeInfo& type_info) const
         {
+            std::lock_guard<std::mutex> guard(get_mutex());
+            return m_op_types.find(type_info) != m_op_types.end();
         }
 
-        template <typename T>
+        /// \brief Return true if OP_TYPE is in the opset
+        template <typename OP_TYPE>
         bool contains_type() const
         {
-            return m_op_types.find(T::type_info) != m_op_types.end();
+            return contains_type(OP_TYPE::type_info);
+        }
+
+        /// \brief Return true if name is in the opset
+        bool contains_type(const std::string& name) const
+        {
+            std::lock_guard<std::mutex> guard(get_mutex());
+            return m_name_type_info_map.find(name) != m_name_type_info_map.end();
         }
 
+        /// \brief Return true if node's type is in the opset
         bool contains_op_type(Node* node) const
         {
+            std::lock_guard<std::mutex> guard(get_mutex());
             return m_op_types.find(node->get_type_info()) != m_op_types.end();
         }
 
     protected:
         std::set<NodeTypeInfo> m_op_types;
+        std::map<std::string, NodeTypeInfo> m_name_type_info_map;
     };
 
-    const OpSet& get_opset0();
-    const OpSet& get_opset1();
+    OpSet& get_opset0();
```

Comment:
Let's return back const reference to OpSet for all default OpSets. If you return non-constant reference to OpSet it means that user could extend default OpSets or (worse than that) replace default operation.
We shouldn't get a chance to change default OpSets after the product release. If user want to add custom operation he should create custom OpSet and don't extend default OpSet.